astropy/nddata/mixins/tests/test_ndio.py

```diff
@@ -12,5 +20,415 @@ class NDDataIO(NDIOMixin, NDData):
+
+        assert ndd3.mask is None
+        assert ndd3.unit is None
+
+    def test_nddata_write_read_wcs_slicing(self):
+        anotherfile = str(self.temp())
+        filename = str(self.temp())
+        ndd1 = NDData(np.ones((10, 10)))
+        write_to_fits(ndd1, filename)
+        ndd2 = read_from_fits(filename)
+        # Need another round trip to compare "new" wcs.
+        # Slice data and wcs to have something real to compare but keep header
+        # and see if it is updated correctly
+        ndd2tmp = NDData(ndd2.data[2:5, 4:8],
+                         wcs=ndd2.wcs[2:5, 4:8], meta=ndd2.meta)
+        write_to_fits(ndd2tmp, anotherfile)
+        ndd3 = read_from_fits(anotherfile)
+
+        np.testing.assert_array_equal(ndd2.data[2:5, 4:8], ndd3.data)
+        assert ndd2.data.dtype == ndd3.data.dtype
+
+        assert ndd3.wcs.wcs.compare(ndd2tmp.wcs.wcs)
+        assert not ndd3.wcs.wcs.compare(ndd2.wcs.wcs)
+
+        assert ndd3.meta != ndd2.meta
+        assert ndd3.meta != ndd2tmp.meta
+
+        assert ndd3.uncertainty is None
+
+        assert ndd3.mask is None
+        assert ndd3.unit is None
+
+    def test_nddata_write_read_meta_wcs(self):
+        filename = str(self.temp())
+        meta = dict([(j, i) for i, j in enumerate('abcdefghijklmnopqrstuvwxyz')])
+        ndd1 = NDData(np.ones((10, 10)), meta=meta)
+        write_to_fits(ndd1, filename)
+        ndd2 = read_from_fits(filename)
+        # Need another round trip to compare "new" wcs.
+        # Slice data and wcs to have something real to compare but keep header
+        # and see if it is updated correctly
+        ndd2tmp = NDData(ndd2.data[2:5, 4:8],
+                         wcs=ndd2.wcs[2:5, 4:8], meta=ndd2.meta)
+
+        anotherfile = str(self.temp())
+        write_to_fits(ndd2tmp, anotherfile)
+        ndd3 = read_from_fits(anotherfile)
+
+        np.testing.assert_array_equal(ndd2.data[2:5, 4:8], ndd3.data)
+        assert ndd2.data.dtype == ndd3.data.dtype
+
+        assert ndd3.wcs.wcs.compare(ndd2tmp.wcs.wcs)
+        assert not ndd3.wcs.wcs.compare(ndd2.wcs.wcs)
+
+        assert ndd3.meta != ndd2.meta
+        assert ndd3.meta != ndd2tmp.meta
+        assert all(ndd1.meta[key] == ndd2.meta[key] for key in ndd1.meta)
+        assert all(ndd1.meta[key] == ndd3.meta[key] for key in ndd1.meta)
+
+        assert ndd3.uncertainty is None
+
+        assert ndd3.mask is None
+        assert ndd3.unit is None
+
+    def test_ndiomixin_read(self):
+        filename = str(self.temp())
+
+        data = np.ones((10, 10), )
+        meta = dict([(j, i) for i, j in enumerate('abcdefghijklmnopqrstuvwxyz')])
+        unit = 'adu'
+        mask = np.random.random((10, 10)) > 0.5
+        uncertainty = UnknownUncertainty(np.ones((5, 5)))
+
+        ndd1 = NDData(data, uncertainty=uncertainty, unit=unit, meta=meta, mask=mask)
+        write_to_fits(ndd1, filename)
+        ndd2 = NDDataIO.read(filename, format='simple_fits')
```

Comment:
@MSeifert04 , my suspicion is that you need to treat `NDDataIO.read()` the same as `fits.read()` here. That is, you need to either:

1. Explicitly close `ndd2` (and any other handlers that used `read()`); OR
2. Use the `read()` in the form of a context manager.